README.md
# change_window

#### Table of Contents

1. [Background](#background)
2. [Overview](#overview)
3. [Module Description](#module-description)
4. [Usage](#usage)
5. [Contributions - Guide for contributing to the module](#contributions)

## Background

This module is a fork of [ipcrm-change_window](https://forge.puppet.com/modules/ipcrm/change_window) which is no longer actively maintained. This fork has been moved under the puppetlabs namespace to allow for community contributions.

It is **not** supported or maintained by puppet.

**Original Author:**
- Matt Cadorette

**Week in Month feature:**
- Steve Jefferies (WTW)

## Overview

Provides change_window functionality that allows you to check current time against change windows and a defined type that applies noop() to classes when not within the change window.


## Module Description

Why?

The original reason for this module was to use it in conjunction with the `trlinkin/noop` module.  However, you can actually use the functions with any resource that you need to be sensitive to change windows by simply wrapping that resource declaration with some conditional logic.

The module is made up of two functions and a defined class.  The functions, [change_window](#change_window)() and [merge_change_windows](#merge_change_windows)(), allow the comparison of a change window schedule against the current time to determine if the run is within the change window ('true'), or outside the change window ('false').  The change_window function checks against a single window definition, while the merge_change_window function will check a list windows and return 'true' if any one of them is true.

The defined type [change_window::apply](#change_windowapply) will accept an array of change windows and a list of classes.  The class list is then included into the catalog with the noop() mode set appropriately.  This define is intended for use during role definition and allows change window control over some or all of the profiles within the role.  Keeping the role definition tidy and allowing some classes to function without change control.  A handy feature when you have changes that you wish applied all the time.

*IMPORTANT:* Remember that if you are _within_ the change window then the value returned by the functions is a String containing _'true'_, otherwise it returns _'false'_.

# Defined types
## change_window::apply
### Usage
```puppet
change_window::apply { 'my_controlled_changes':
  change_window_set => hiera('weekly_window'),
  class_list        => [ 'profile::ntp', 'profile::resolver' ],
}
```
where:

`change_window_set` = An array of arrays defining your change windows to check.  Most easily defined and retrieved via hiera (but does not have to be, see [merge_change_windows](#merge_change_windows) for details)

`class_list`        = an array of classes to include


The class_list will accept either simple class names to include or a hash describing the class/resource along with its parameters.

example class_list: with simple and parameterized classes
```
$class_list = [
  'profile::parameter_class' => {
    parm1 => 'value1',
    parm2 => 'value2',
  },
  'profile::simple_class',
]
```

# Functions
## change_window
### Usage
change_window( $tz, $window_type, $window_wday, $window_time, [$window_week], [$window_month], [$time])

Where:

- `$tz` is the timezone offset you want used when the current timestamp is generated.(this example is for EST)

- `$window_type` accepts to values: per_day or window.
  -  `per_day` tells change_window that the hours specified are valid on each day specified.  For example if you set days 0-3 and start 20:00, end 23:00 - then Sunday through Wednesday from 8PM to 11PM this function will return true. For wrapped hours you receive start to midnight on the first day and midnight to end on the last day.
  -   `window` tells change_window to treat the days and times as a continuous change window spanning from start day/start time through end day/end time.

- `$window_wday` is a hash where start is the first weekday in your window and end is the last weekday - expressed as weekday names or 0-6.  You can specify the same day if you like and you may wrap the weekend (i.e friday .. monday).

- `$window_time` is a hash where the start key is a timestamp (HH:MM), and end sets the end hour and minute. You may wrap the midnight hour (i.e. 22:00 .. 02:00). For per_day windows that wrap the midnight hour, the first day will apply the start-to-midnight and the last day will apply the midnight-to-end of the window.
- `$window_week` **OPTIONAL** array of weeks within a month to accept as within the change window. Values in the array must be of range 1-6. See [Week in month](#week-in-month-change-windows) for details.
- `$window_month` **OPTIONAL** array of months within a year to accept as within the change window. Values in the array must be of range 1-12. See [Month in year](#month-in-year-change-windows) for details.
- `$time` is an optional parameter that lets you specify the time to test as an array.  This array is passed to the Time.new() object to set the time under test.  This array should take the form of [ YYYY, MM, DD, HH, MM] and will apply the timezone specified.


```puppet
$tz = "-05:00"
$window_wday  = { start => 'Friday', end => 'Saturday' }
$window_time = { start  => '20:00', end => '23:00' }
$window_type = 'window'
$val = change_window($tz, $window_type, $window_wday, $window_time)

if $val == 'false' {
    notify { "Puppet noop enabled in site.pp! Not within change window!": }
    noop()
}
```

Another example shows wrapping the weekend.  You can specify combinations like below (days 5 - 0).  This will result in days 5,6,0 as being valid for the change window.  In this case I'm using per_day so on Friday, Saturday, or Sunday between 8PM and 11PM changes will be allowed.

```puppet
$tz = "-05:00"
$window_wday  = { start => 'Friday', end => 'Sunday' }
$window_time = { start  => '20:00', end => '23:00' }
$window_type = 'per_day'
$val = change_window($tz, $window_type, $window_wday, $window_time)

if $val == 'false' {
    notify { "Puppet noop enabled in site.pp! Not within change window!": }
    noop()
}
```

Example of using week and month windows, run on first and second week of the month and only between January and March.

```puppet
$tz = "-05:00"
$window_wday  = { start => 'Friday', end => 'Saturday' }
$window_time = { start  => '20:00', end => '23:00' }
$window_type = 'window'
$window_week = [1,2]
$window_month = [1,2,3]
$val = change_window($tz, $window_type, $window_wday, $window_time, $window_week, $window_month)

if $val == 'false' {
    notify { "Puppet noop enabled in site.pp! Not within change window!": }
    noop()
}
```

You can also use hiera to enable more complex windows:

hiera:
```yaml
tz_dev: "-05:00"
window_type_dev: per_day
window_wday_dev:
  start: Friday
  end: Sunday
window_time_dev:
  start: "20:00"
  end: "23:00"
```

site.pp:
```puppet
$e = $::custom_env
$val = change_window(
         hiera("tz_${e}"),
         hiera("window_type_${e}"),
         hiera("window_wday_${e}"),
         hiera("window_time_${e}")
         )
if $val == 'false' {
  notify { "Puppet noop enabled in site.pp for env ${e}! Not within change window!": }
  noop()
}
```

## merge_change_windows
### Usage
merge_change_windows( $list_of_windows )

Where:

$list_of_windows is an Array made up of Arrays containing change_window parameters.

#### Manifest Example
```puppet
$tz = "-05:00"

# Friday @ 10 PM until Monday @ 2 AM
$window1_type = 'window'
$window1_wday = { start => 'Friday', end => 'Monday' }
$window1_time = { start => '22:00',  end => '02:00' }

# Wednesday @ 10 PM until Thursday @ 2 AM
$window2_type = 'window'
$window2_wday = { start => 'Wednesday', end => 'Thursday' }
$window2_time = { start => '22:00',     end => '02:00' }

$change_windows = [
  [$tz, $window1_type, $window1_wday, $window1_time],
  [$tz, $window2_type, $window2_wday, $window2_time],
]

if merge_change_windows($change_windows) == 'false' {
  notify { "Puppet noop enabled in site.pp! Not within change window!": }
  noop()
}
```

#### Hiera Example
hiera:
```yaml
change_window_set::my_change_window:
  - # Friday @ 10 PM until Monday @ 2 AM
    - '-05:00'
    - 'window'
    - start: 'Friday'
      end:   'Monday'
    - start: '22:00'
      end:   '02:00'
  - # Wednesday @ 10 PM until Thursday @ 2 AM
    - '-05:00'
    - 'window'
    - start: 'Wednesday'
      end:   'Thursday'
    - start: '22:00'
      end:   '02:00'
```

site.pp:
```puppet
$change_window_set = 'my_change_window'
$change_windows    = hiera("change_window_set::${my_change_window}")

if merge_change_windows($change_windows) == 'false' {
  notify { "Puppet noop enabled in site.pp! Not within change window!": }
  noop()
}
```

### Week in Month Change Windows
Using the `$window_week` parameter in `change_window` you can specify a sub-set of weeks within each month to be 
accepted as valid change windows.

This parameter is optional, if not used all weeks are included

The parameter takes an Array of integers, between 1-6 as valid week definitions.

#### Example

Given the month is April 2021:
![window week example](img/window_week.png?raw=true "Window week example - April 2021")

Then:
- Week 1 runs from first of the month to the first Sunday
- Week 2 runs from the following Monday to the next Sunday
- etc...
- Last week of the month runs from final Monday to last day of the month

### Month in Year Change Windows
Using the `$window_month` parameter in `change_window` you can specify a sub-set of months within each year to be 
accepted as valid change windows.

This parameter is optional, if not used all months are included.

The parameter takes an array of integers and/or string ranges of values.

Months are represented by numerical values:

| Month     | Value |
| -----     | ----- |
| January   | 1     |
| February  | 2     |
| March     | 3     |
| April     | 4     |
| May       | 5     |
| June      | 6     |
| July      | 7     |
| August    | 8     |
| September | 9     |
| October   | 10    |
| November  | 11    |
| December  | 12    |

For example:
- Run only every other month: `[1,3,5,7,9,11]`
- Run only between January and March (inclusive): `[1,2,3]` or `['1-3']`

Individual values and ranges must only be between 1 and 12.

## Acknowledgements

- window_week functionality is provided through an extension which is a modified version of
  [week-of-month](https://github.com/sachin87/week-of-month)

# Contributions  

If anyone would like to contribute to the module, PR's are welcome.

If you're experiencing any bugs you can raise an issue below however, given that this is a community module, it will only be resolved if someone submits a PR with a fix. Puppet accepts no responsibility for this module as described in [Background](#background).

Repo:        https://github.com/puppetlabs/ipcrm-change_window

Issues link: https://github.com/puppetlabs/ipcrm-change_window/issues

CHANGELOG
2016-01-28 Release 0.1.5
- merge_change_windows function added
2016-01-27 Release 0.1.4
- Merged PR#1, see notes
2016-01-25 Release 0.1.0
- Init


spec/spec_helper.rb
require 'puppetlabs_spec_helper/module_spec_helper'
require 'rspec-puppet-facts'

include RspecPuppetFacts

require 'simplecov'
require 'simplecov-console'

RSpec.configure do |c|
  c.module_path  = File.expand_path(File.join(__FILE__, '../fixtures/modules'))
  c.manifest_dir = File.expand_path(File.join(__FILE__, '../fixtures/manifests'))
  c.hiera_config = File.expand_path(File.join(__FILE__, '../fixtures/hiera.yaml'))
  c.before(:all) do
    if ENV['DEBUG']
      Puppet::Util::Log.level = :debug
      Puppet::Util::Log.newdestination(:console)
    end
  end
end



SimpleCov.start do
  add_filter '/spec'
  add_filter '/vendor'
  formatter SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::HTMLFormatter,
    SimpleCov::Formatter::Console
  ])
end


spec/classes/test_apply_spec.rb
require 'spec_helper'

describe 'test_apply', :type => :class do
  context 'on all OSes' do
    let :facts do
      {
        :id                     => 'root',
        :kernel                 => 'Linux',
        :osfamily               => 'RedHat',
        :operatingsystem        => 'RedHat',
        :operatingsystemrelease => '6',
        :concat_basedir         => '/dne',
        :path                   => '/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin',
        :is_pe                  => false,
      }
    end
    it { should contain_class("test_notify_simple") }
    it { should contain_class("test_notify_parameter") }
    it { should contain_notify("test_notify_parameter").with_noop(true)}
    it { should contain_notify("test_notify_simple").without_noop}
  end
end


spec/functions/change_window_spec.rb
require 'spec_helper'
tz = "-05:00"
time  = [2016,1,6,6,15] # 2016-01-06 06:15 (Wed)
oTime = Time.new( *time, 0, tz)

Puppet::Util::Log.level = :debug
Puppet::Util::Log.newdestination(:console)

# Bad window_time
fail_wndw_type = 'windows'
fail_key_day   = {'starts' => '00:00', 'ends' => '23:59'}
fail_key_time  = {'starts' => '00:00', 'ends' => '23:59'}

# Build some test dates
pass_wndw_day  = {'start' => 2, 'end' => 4} # Window includes
fail_wndw_day  = {'start' => 1, 'end' => 2} # Window excludes
pass_one_day   = {'start' => 3, 'end' => 3} # One day includes
fail_one_day   = {'start' => 1, 'end' => 1} # One day excludes
pass_start_day = {'start' => 3, 'end' => 5} # Window starts on t
pass_end_day   = {'start' => 2, 'end' => 3} # Window ends on t
pass_wrap_day  = {'start' => 6, 'end' => 4} # Weekend wrap includes
fail_wrap_day  = {'start' => 6, 'end' => 2} # Weekend wrap excludes
pass_all_days  = {'start' => 0, 'end' => 6} # All days valid
pass_string_days = {'start' => 'sUn', 'end' => 'SATurday'}
fail_string_days = {'start' => 'today', 'end' => 'tommorow'}

# Build some test times
pass_wndw_time       = {'start' => '06:15', 'end' => '06:30'} # Window includes
fail_wndw_time       = {'start' => '06:00', 'end' => '06:14'} # Window excludes
pass_wrap_time_start = {'start' => '06:00', 'end' => '05:00'} # Wrap includes start < t
pass_wrap_time_end   = {'start' => '11:00', 'end' => '06:15'} # Wrap includes end > t
pass_all_times       = {'start' => '00:00', 'end' => '23:59'} # All times valid
fail_time_format     = {'start' => '0000', 'end' => '2359'}   # Times missing colon (:)

describe 'change_window' do
  # Test for parseError's
  it { is_expected.not_to eq(nil) }
  it { is_expected.to run.with_params().and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(1).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz, fail_wndw_type, pass_all_days, pass_wndw_time).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz,'window', pass_all_days,pass_all_times, [2016,1,6,6,60]).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz,'window', fail_key_day,pass_all_times, time).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz,'window', pass_all_days,fail_key_time).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz,'window', pass_all_days,fail_time_format).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz,'window', fail_string_days,pass_all_times).and_raise_error(Puppet::ParseError) }
  it { is_expected.to run.with_params(tz,'window', pass_string_days,pass_all_times, time).and_return(true) }
  # Test day-of-week (window)
  it { is_expected.to run.with_params(tz, 'window', pass_wndw_day,  pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', fail_wndw_day,  pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'window', pass_one_day,   pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', fail_one_day,   pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'window', pass_start_day, pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', pass_end_day,   pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', fail_one_day,   pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'window', pass_wrap_day,  pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', fail_wrap_day,  pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'window', pass_start_day, pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', pass_end_day,   pass_all_times, time).and_return(true) }
  # Test day-of-week (per_day)
  it { is_expected.to run.with_params(tz, 'per_day', pass_wndw_day,  pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', fail_wndw_day,  pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_one_day,   pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', fail_one_day,   pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_start_day, pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_end_day,   pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', fail_one_day,   pass_all_times, time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_wrap_day,  pass_all_times, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', fail_wrap_day,  pass_all_times, time).and_return(false) }
  # Test time-of-day (window)
  it { is_expected.to run.with_params(tz, 'window', pass_one_day,   pass_wndw_time,       time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', pass_one_day,   fail_wndw_time,       time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'window', pass_wndw_day,  pass_wndw_time,       time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', pass_wndw_day,  fail_wndw_time,       time).and_return(true) } # passes because mid-window
  it { is_expected.to run.with_params(tz, 'window', pass_start_day, pass_wrap_time_start, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'window', pass_end_day,   pass_wrap_time_end,   time).and_return(true) }
  # Test time-of-day (per_day)
  it { is_expected.to run.with_params(tz, 'per_day', pass_one_day,   pass_wndw_time,       time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_one_day,   fail_wndw_time,       time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_wndw_day,  pass_wndw_time,       time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_wndw_day,  fail_wndw_time,       time).and_return(false) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_start_day, pass_wrap_time_start, time).and_return(true) }
  it { is_expected.to run.with_params(tz, 'per_day', pass_end_day,   pass_wrap_time_end,   time).and_return(true) }
end


lib/puppet/parser/functions/merge_change_windows.rb
require 'date'

module Puppet::Parser::Functions
  newfunction(
    :merge_change_windows,
    :doc  => "Creates complex change windows by merging a series of windows together",
    :type => :rvalue
  ) do |args|

    # Validate arguments
    raise Puppet::ParseError, "Invalid argument count, got #{args.length} and expected 1" unless args.length == 1
    raise Puppet::ParseError, "Change_windows must be an array!" unless args[0].is_a? Array

    in_cw = false
    args[0].each { |cw|
      raise Puppet::ParseError, "Expect an Array for change_window entry, received #{cw.class}" unless cw.is_a? Array
      begin
        if function_change_window( cw ) == true
          in_cw = true
        end
      rescue Exception => e
        # Catch exception and rebrand it as ours
        raise Puppet::ParseError, "change_window threw #{e.message}"
      end
    }
    return in_cw
  end
end


lib/puppet/parser/functions/change_window.rb
require 'date'

module Puppet::Parser::Functions
  newfunction(
    :change_window,
    :doc  => "Provides change_window function that allows you to check current time against change window",
    :type => :rvalue
  ) do |args|
    class << self
      def change_window_time_is_within(s,e,c)
        # Calculate by minutes of the week 0 ... 1439
        test = ((s[0].to_i*60+s[1].to_i)..(e[0].to_i*60+e[1])).cover?(c[0]*60+c[1])
        return test
      end
    end

    # args[0] TimeZone
    # args[1] Window Type - per_day or window(default)
    # args[2] Hash, window_wday
    # args[3] Hash, window_time
    # args[4] String, Point-in-time to test (optional)

    # Validate Arguments are all present
    raise Puppet::ParseError, "Invalid argument count, got #{args.length} expected 4 or 5" unless (4..5).cover?(args.length)

    # Validate Args are correct type
    raise Puppet::ParseError, "TimeZone must be a string!"    unless args[0].is_a? String
    raise Puppet::ParseError, "Window type must be a string!" unless args[1].is_a? String
    raise Puppet::ParseError, "Window_wday must be a hash!"   unless args[2].is_a? Hash
    raise Puppet::ParseError, "Window_time must be a hash!"   unless args[3].is_a? Hash
    if( args.length == 5)
      raise Puppet::ParseError, "Point-in-time must be an array!" unless args[4].is_a? Array
    end

    # Set Timezone for other timestamp
    timezone = args[0]

    # Evaluation Type
    # per_day, or window
    # If per_day, we'll expect that window exists between start and end for each included
    # day.
    # If window, we'll expect that the window is continuous - from start day, start time through end day, end time.
    window_type = args[1].nil? ? 'window' : args[1]

    if !['window','per_day'].include?(window_type)
      raise Puppet::ParseError, "Window type must be 'window' or 'per_day'"
    end

    ## Setup the days of the week hash
    # First, validate that we got the keys we expect
    args[2].keys.each { |ww|
      if not ['start','end'].include?(ww)
        raise Puppet::ParseError, "The window_wday hash can include start and end only! Invalid value passed - #{ww}"
      end
    }

    # Now that we now we have valid keys, set local var equal to passed
    window_wday  = args[2]

    # Figure out what days have been passed & Convert non-int
    # Create new hash for storing ONLY int based days
    window_wday_int = Hash.new

    # For each day passed in, conver it from text to numeric
    # if numeric wasn't used
    for wdkey in window_wday.keys
      if window_wday[wdkey].is_a? String
        window_wday_int[wdkey] = Date::ABBR_DAYNAMES.map(&:downcase).find_index(window_wday[wdkey][0..2].downcase)
      elsif window_wday[wdkey].is_a? Integer
        window_wday_int[wdkey] = window_wday[wdkey]
      else
        raise Puppet::ParseError, "Invalid day found in supplied window days(its not a string or integer)"
      end
    end

    # One more validation - did we lookup valid values?
    window_wday_int.keys.each { |wi|
      if !window_wday_int[wi].is_a? Integer
        raise Puppet::ParseError, "Invalid day found in supplied window days"
      end

      if !(0..6).cover?(window_wday_int[wi])
        raise Puppet::ParseError, "Supplied weekday number is out of range (i.e. not 0-6) - value found #{window_wday_int[wi]}"
      end
    }

    ## Stores window start end time
    #24Hour #Minute
    #Example: {'start' => '17:00', 'end' => '23:00'}

    # Validate we got a usable time hash provided
    args[3].keys.each { |ts|
      if not ['start','end'].include?(ts)
        raise Puppet::ParseError, "Invalid key provided for window_time. Only start/end supported - found #{ts}"
      end

      if not args[3][ts] =~ /^\d(\d)?\:\d\d$/
        raise Puppet::ParseError, "Invalid time supplied for window_time #{ts} - found #{args[3][ts]}"
      end
    }

    # Now that we know its valid store it as a hash
    start_time = args[3]['start'].split(':').map(&:to_i)
    end_time   = args[3]['end'].split(':').map(&:to_i)
    window_time = {'start' => start_time, 'end' => end_time }
    window_time_str = {'start' => sprintf('%02d:%02d', start_time[0], start_time[1]),
      'end' => sprintf('%02d:%02d',end_time[0],end_time[1])}

    # Get Time (this will use localtime)
    if args.length != 5
      t = Time.now.getlocal(timezone)
    else
      raise Puppet::ParseError, "Invalid key for time expected 5 values and received #{args[4].length}" unless args[4].length == 5
      begin
        t = Time.new( *args[4], 0, timezone)
      rescue Exception
        # Catch exception and rebrand as parse error
        raise Puppet::ParseError, "Could not convert time array into valid Time.new object, received #{args[4].to_a}"
      end
    end

    # Build array of valid_days
    if window_wday_int['start'] > window_wday_int['end']
      # EOW-Wrap 4,1 = 4,5,6,0,1
      valid_days = (window_wday_int['start']..6).to_a + (0..window_wday_int['end']).to_a.uniq
    else
      # Within-EOW 1,4 = 1,2,3,4
      valid_days = (window_wday_int['start']..window_wday_int['end']).to_a.uniq
    end

    # Determine if today is within the valid days for the change window
    if valid_days.include?(t.wday)

      # IF this is the first day of the window
      if t.wday == window_wday_int['start'] and valid_days.length > 1
        # If window type or per_day with midnight wrap
        if window_type == 'window' or (window_type == 'per_day' and window_time_str['start'] > window_time_str['end'])
          # IF we are within <start time> and 23:59 return true
          window_time['end'] = [23,59]
        end
        return change_window_time_is_within(window_time['start'],window_time['end'],[t.hour,t.min]) == true ? true : false

      # If this is the last day of the window
      elsif t.wday == window_wday_int['end'] and valid_days.length > 1

        # If window type or per_day with midnight wrap
        if window_type == 'window' or (window_type == 'per_day' and window_time_str['start'] > window_time_str['end'])
          #If we are within 00:00 and end
          window_time['start'] = [0, 0]
        end
        return change_window_time_is_within(window_time['start'],window_time['end'],[t.hour,t.min]) == true ? true : false

      # If 1 day window type or per_day wo/ midnight wrap
      elsif (valid_days.length == 1 and window_type == 'window') or window_type == 'per_day'
        return change_window_time_is_within(window_time['start'],window_time['end'],[t.hour,t.min]) == true ? true : false


      # midweek per_day
      elsif window_type == 'per_day'
        return change_window_time_is_within(window_time['start'],window_time['end'],[t.hour,t.min]) == true ? true : false

      # Fall through matches window type and between window start/end
      else
        return true
      end

    # Your not within the valid_days
    else
      return false
    end
  end
end
